File: day10/Day10.scala
import scala.io.Source.stdin

val starting = "([{<"
val ending = ")]}>"

@main def main(idx: Int) =
  val first: Either[Char, List[Char]] = Right(List[Char]())
  val xs = stdin.getLines().map{
             _.foldLeft{first}{(r, c) => r match {
               case Right(l) if l.isEmpty || starting.contains(c) => Right(c :: l)
               case Right(prev :: l) => if prev == starting(ending.indexOf(c))
                                        then Right(l)
                                        else Left(c)
               case r => r
             }}}
  if idx == 1 then
    val scores = Map(')' -> 3, ']' -> 57, '}' -> 1197, '>' -> 25137)
    print{xs.flatMap(_.left.toOption).map(scores).sum}
  else
    val result = xs.flatMap(_.toOption).map{_.map{c => starting.indexOf(c) + 1}
                                             .foldLeft{0L}((a, b) => a * 5 + b)}
                                       .toIndexedSeq.sorted
    print{result(result.size / 2)}

File: day10/Day10.rs
use std::env;
use std::io;

const STARTING: &str = "([{<";
const ENDING: &str = ")]}>";
fn main() {
    let matching =
        |src: &str, c: char, target: &str| src.find(c).and_then(|i| target.chars().nth(i));

    type Partial = Vec<char>;
    type Wrong = char;

    let mut parse_results: Vec<Result<Partial, Wrong>> = Vec::new();

    for l in io::stdin().lines().map(Result::unwrap) {
        let mut wrong = false;
        let mut stack: Partial = Vec::with_capacity(l.len());

        for c in l.chars() {
            if STARTING.contains(c) || stack.is_empty() {
                stack.push(c)
            } else if let Some(prev) = stack.pop() {
                if prev != matching(ENDING, c, STARTING).unwrap() {
                    parse_results.push(Err(c));
                    wrong = true;
                    break;
                }
            }
        }

        if !wrong {
            parse_results.push(Ok(stack));
        }
    }

    if env::args().nth(1).unwrap() == "1" {
        let result: u32 = parse_results
            .iter()
            .filter_map(|r| r.as_ref().err())
            .map(|c| match c {
                ')' => 3,
                ']' => 57,
                '}' => 1197,
                '>' => 25137,
                _ => 0,
            })
            .sum();

        println!("{}", result);
    } else {
        let mut scores: Vec<usize> = parse_results
            .iter()
            .filter_map(|v| v.as_ref().ok())
            .map(|v| {
                v.iter()
                    .rev()
                    .map(|c| matching(STARTING, *c, ENDING).unwrap())
                    .map(|c| ENDING.find(c).unwrap() + 1)
                    .fold(0, |a, b| a * 5 + b)
            })
            .collect();

        scores.sort_unstable();

        println!("{}", scores[scores.len() / 2]);
    }
}
